unleash-yggdrasil/src/lib.rs

```diff
@@ -94,84 +158,144 @@ impl EngineState {
         }
     }
 
-    fn enabled(&self, toggle: Option<&CompiledToggle>, context: &Context) -> bool {
-        if let Some(toggle) = toggle {
-            let context = EnrichedContext::from(context.clone(), toggle.name.clone());
-            toggle.enabled && (toggle.compiled_strategy)(&context)
+    fn harvest_metrics(&mut self) -> Option<MetricBucket> {
+        if let Some(state) = &self.compiled_state {
+            let metrics: HashMap<String, ToggleStats> = state
+                .values()
+                .map(|toggle| {
+                    let yes = toggle.yes.swap(0, Ordering::Relaxed);
+                    let no = toggle.no.swap(0, Ordering::Relaxed);
+
+                    let mut variants: HashMap<String, u32> = toggle
+                        .variants
+                        .iter()
+                        .map(|variant| {
+                            (
+                                variant.name.clone(),
+                                variant.count.swap(0, Ordering::Relaxed),
+                            )
+                        })
+                        .collect();
+
+                    variants.insert(
+                        VariantDef::default().name,
+                        toggle.default_variant.swap(0, Ordering::Relaxed),
+                    );
+
+                    (toggle.name.clone(), ToggleStats { yes, no, variants })
+                })
+                .collect();
+            let timestamp = Utc::now();
+            let bucket = MetricBucket {
+                toggles: metrics,
+                start: self.started,
+                stop: timestamp,
+            };
+            self.started = timestamp;
+            Some(bucket)
+        } else {
+            None
+        }
+    }
+
+    pub fn get_metrics(&mut self) -> Option<MetricBucket> {
+        self.harvest_metrics()
+    }
+
+    fn enabled(&self, toggle: &CompiledToggle, context: &Context) -> bool {
+        let context = EnrichedContext::from(context.clone(), toggle.name.clone());
+        let is_enabled = toggle.enabled && (toggle.compiled_strategy)(&context);
+        if is_enabled {
+            toggle.yes.fetch_add(1, Ordering::Relaxed);
         } else {
-            false
+            toggle.no.fetch_add(1, Ordering::Relaxed);
         }
+        is_enabled
     }
 
     pub fn is_enabled(&self, name: String, context: &Context) -> bool {
-        match &self.compiled_state {
-            Some(_) => {
-                let toggle = self.get_toggle(name);
-                self.enabled(toggle, context)
+        self.compiled_state
+            .as_ref()
+            .and_then(|_| {
+                self.get_toggle(name)
+                    .map(|toggle| self.enabled(toggle, context))
+            })
+            .unwrap_or_default()
+    }
+
+    fn resolve_variant<'a>(
+        &self,
+        toggle: &CompiledToggle,
+        variants: &'a Vec<CompiledVariant>,
+        context: &Context,
+    ) -> Option<&'a CompiledVariant> {
```

Comment:
From a quick search does `'a` here mean that our result type only lives as long as what we're taking in as our parameters? Would love a ELI5 in this context if possible, thanks!

Reply:
Your Google-Fu is most excellent. Right let's unpack this beast. 

In our function signature we declare `fn resolve_variant<'a>` which tells Rust that this function is going to receive some borrowed data and that the return value is _also_ going to be a borrow but it'll be borrowed against the incoming data. The 'a here is telling Rust that we're going to call this lifetime `'a`.

In the parameters we declare `variants: &'a Vec<CompiledVariant>` which tells Rust that this parameter has a lifetime of `'a` - this is the parameter that fulfils the lifetime declaration in the function signature and tells Rust that the lifetime of `'a` is as long as variants lives.

In the return value we have `Option<&'a CompiledVariant>` so now we're telling Rust that the return value must live at least as long as whatever data is bound to the `'a` lifetime.

So in English: we're borrowing a chunk of data called variants, when we return the data from this function, the data in variants can't be discarded before the return value is.

Reply:
That's perfect and really helpful, thanks Simon! 
@gastonfournier you may find this insightful as well.